Build failing when using meson 1.4.0rc1

In order to test the impact of Meson 1.4.0 , I upgraded a test repository of openSUSE Tumbleweed to this new version,

Out of the builds, umockdev is failing (seems its failing similar to what that HACK commit addresses, which was done in Gentoo? Did Gentoo update to a 1.4 pre-release of meson?)

I get plenty of these errors:

```
[   12s] umockdev-record.p/src/umockdev-record.c:1141:22: error: unused variable ‘_line_collection_size_’ [-Werror=unused-variable]
[   12s]  1141 |                 gint _line_collection_size_ = 0;
[   12s]       |                      ^~~~~~~~~~~~~~~~~~~~~~
```

Simply going back to meson 1.3.2 makes umockdev build pass (CC @jpakkane / meson release maintainer)

<strike>(full build log can be found at https://build.opensuse.org/build/GNOME:Next/openSUSE_Factory/x86_64/umockdev/_log )</strike>

--- a/meson.build
+++ b/meson.build
@@ -21,7 +21,6 @@ add_project_arguments(
   '-Werror=format-security',
   '-Werror=format=2',
   #'-Werror=missing-include-dirs',
-  '-Werror=unused-variable',
   '-Werror=return-type',
   '-Werror=uninitialized',
   # experimental: first check this on all Debian architectures
@@ -32,30 +31,23 @@ add_project_arguments(
 
 add_project_arguments('--abi-stability', language: 'vala')
 
-# let's not clutter the code with too many #ifdefs
-if get_option('b_ndebug') == 'true'
-  add_project_arguments('-Wno-error=unused-but-set-variable', '-Wno-error=unused-variable', language: 'c')
-endif
-
-# HACK: between Jan 5 and Jan 8 2024, Gentoo changed something that causes
-# hundreds of warnings when compiling the vala-generated C code; ignore these
-if get_option('gentoo-hacks')
-  add_project_arguments(
-    '-Wno-error',
-    '-Wno-error=incompatible-pointer-types',
-    '-Wno-error=pointer-sign',
-    '-Wno-error=unused-function',
-    '-Wno-error=unused-label',
-    '-Wno-error=unused-but-set-variable',
-    '-Wno-error=unused-variable',
-    '-Wno-error=unused-value',
-    language: 'c')
-endif
-
 conf = configuration_data()
 cc = meson.get_compiler('c')
 valac = meson.get_compiler('vala')
 g_ir_compiler = find_program('g-ir-compiler', required: false)
+
+# generated vala code triggers tons of these
+add_project_arguments(
+  cc.get_supported_arguments(
+    '-Wno-error=discarded-qualifiers',
+    '-Wno-error=incompatible-pointer-types',
+    '-Wno-error=pointer-sign',
+    '-Wno-error=unused-but-set-variable',
+    '-Wno-error=unused-function',
+    '-Wno-error=unused-label',
+    '-Wno-error=unused-value',
+    '-Wno-error=unused-variable'
+  ), language: 'c')
 
 # HACK: vala doesn't support declaring const arguments, so we get a lot of errors
 if cc.get_id() == 'clang'

--- a/meson_options.txt
+++ b/meson_options.txt
@@ -2,8 +2,3 @@ option('gtk_doc',
        type : 'boolean',
        value : false,
        description : 'use gtk-doc to build documentation')
-
-option('gentoo-hacks',
-       type : 'boolean',
-       value : false,
-       description : 'disable gcc fatal warnings, for Gentoo builds')

--- a/src/umockdev-record.vala
+++ b/src/umockdev-record.vala
@@ -61,7 +61,7 @@ all_devices ()
 {
     var devs = new GenericArray<string>();
     devices_from_dir("/sys/devices", ref devs);
-    return devs.data;
+    return devs.steal();
 }
 
 // If dev is a block or character device, convert it to a sysfs path.

--- a/tests/run-gentoo
+++ b/tests/run-gentoo
@@ -27,7 +27,7 @@ git config --global safe.directory /source
 
 cd /source
 export VALAC=\$(ls /usr/bin/valac-* |sort | tail -n1)
-meson setup /tmp/dbg --buildtype debug --werror -Dgentoo-hacks=true
+meson setup /tmp/dbg --buildtype debug --werror
 
 export BRITTLE_TESTS="${BRITTLE_TESTS:-}"
 meson test -C /tmp/dbg -v --num-processes=1

